File: algorithm/leetcode1004_max-consecutive-ones-iii.cpp
class Solution {
public:
    int longestOnes(vector<int>& A, int K) {
        int left = 0, right, maxlen = 0;
        for(right = 0; right < A.size(); ++right) 
        {
        	if(A[right]==0)
        	{
        		if(K == 0)
        		{
        			while(A[left] == 1)//等于0退出
        				++left;
        			++left;//退出了一个0
        		}
        		else//k不为0，0变成1的数量还没满
	        	{
	        		--K;
	        	}
        	}	
        	maxlen = max(maxlen,right-left+1);
        }
        return maxlen;
    }
};

File: algorithm/Python3_version/leetcode1004.max-consecutive-ones-iii.py
class Solution:
    def longestOnes(self, A: List[int], K: int) -> int:
        l, r = 0, 0
        maxlen = 0
        while r < len(A):
            if A[r]==0:
                K -= 1
            while K < 0:
                if A[l]==0:
                    K += 1
                l += 1
            r += 1
            maxlen = max(maxlen, r-l)    
        return maxlen